File: 1253-sort-the-matrix-diagonally/sort-the-matrix-diagonally.py
# Optimal Code
class Solution:
    # sort every diagonal of the matrix in ascending order                    
    def diagonalSort(self, mat: List[List[int]]) -> List[List[int]]:
        # number of rows and columns
        m = len(mat)
        n = len(mat[0])
        # dicitonary to map : (row --> col) -> element of that diagonal
        mp = {}
        # collect all element of each diagonal
        for i in range(m):
            for j in range(n):
                # cells with the same values  of (row - columns) belong to the same diagonal
                key = i - j
                # crate the new list for this diagonal if it does not exists
                if key not in mp:
                    mp[key] = []
                # store the current element
                mp[key].append(mat[i][j])
        # sort the element of every diagonal
        for key in mp:
            mp[key].sort()
        # sorted values back into the matrix traverse from bottom right so that pop() removes the smallest remainning element from each sorted diagonal 
        for i in range(m - 1, -1, -1):
            for j in range(n - 1, -1, -1):
                key = i - j
                mat[i][j] = mp[key].pop()
        # return the matrix with sorted diagonals
        return mat

# Time Complexity : O(Nlog)
# Space Complexity : O(N)

File: 1253-sort-the-matrix-diagonally/sort-the-matrix-diagonally.cpp
// Brute Force Code
class Solution {
public:
    // sort every diagonal of the matrix in ascending order
    vector<vector<int>> diagonalSort(vector<vector<int>>& mat) {
        // number of rows and columns
        int rows = mat.size();
        int cols = mat[0].size();
        // function to sort a single diagonal starting from (startRow, startCol)
        auto sortDiagonal = [&](int startRow, int startCol) {
            // store all elements of the current diagonal
            vector<int> diagonal;
            int row = startRow;
            int col = startCol;
            // collect the diagonal elements
            while (row < rows && col < cols) {
                diagonal.push_back(mat[row][col]);
                row++;
                col++;
            }
            // sort the collected elements
            sort(diagonal.begin(), diagonal.end());
            // put the sorted elements back into the same diagonal
            row = startRow;
            col = startCol;
            int index = 0;
            while (row < rows && col < cols) {
                mat[row][col] = diagonal[index];
                index++;
                row++;
                col++;
            }
        };
        // sort all diagonals that start from the first row
        for (int col = 0; col < cols; col++) {
            sortDiagonal(0, col);
        }
        // sort all diagonals that start from the first column skip the top-left cell because its diagonal has already been processed
        for (int row = 1; row < rows; row++) {
            sortDiagonal(row, 0);
        }
        // return the matrix with all diagonals sorted
        return mat;
    }
};

// Time Complexity : O(N)
// Space Complexity : O(N)